web/packages/teleterm/src/mainProcess/runtimeSettings.ts

```diff
@@ -218,21 +218,67 @@ function requestGrpcServerAddresses(): GrpcServerAddresses {
     }
     case 'linux':
     case 'darwin':
-      return {
-        tsh: getUnixSocketNetworkAddress('tsh.socket'),
-        shared: getUnixSocketNetworkAddress('shared.socket'),
-        tshdEvents: getUnixSocketNetworkAddress('tshd_events.socket'),
-      };
+      // Sockets are created in a temporary directory to avoid exceeding the
+      // maximum allowed path length for Unix domain sockets.
+      // Limits: macOS - 104 characters, Linux - 108 characters.
+      const tempDir = await getUnixUserTempDir();
+      const [tsh, shared, tshdEvents] = await Promise.all([
+        getUnixSocketNetworkAddress(tempDir, 'tsh.socket'),
+        getUnixSocketNetworkAddress(tempDir, 'shared.socket'),
+        getUnixSocketNetworkAddress(tempDir, 'tshd_events.socket'),
+      ]);
+      return { tsh, shared, tshdEvents };
   }
 }
 
-function getUnixSocketNetworkAddress(socketName: string) {
-  const unixSocketPath = path.resolve(app.getPath('userData'), socketName);
+async function getUnixSocketNetworkAddress(
+  tempDir: string,
+  socketName: string
+) {
+  const socketPath = path.join(tempDir, socketName);
+  try {
+    // Try to clean up after a previous process that unexpectedly crashed.
+    await fs.promises.unlink(socketPath);
+  } catch (error) {
+    if (error.code !== 'ENOENT') {
+      throw error;
+    }
+  }
+
+  return `unix://${socketPath}`;
+}
 
-  // try to cleanup after previous process that unexpectedly crashed
-  if (fs.existsSync(unixSocketPath)) {
-    fs.unlinkSync(unixSocketPath);
+async function getUnixUserTempDir(): Promise<string> {
+  const appTempDir = app.getPath('temp');
+  const appName = app.getName();
+  // On macOS, 'temp' dir is like //var/folders/y5/yqg8xz555_v7xfsr0wn6b4q80000gn/T/.
+  // That dir is per-user and secure by default, so just create the app subdirectory.
+  if (process.platform === 'darwin') {
+    const dirPath = path.join(appTempDir, appName);
+    await fs.promises.mkdir(dirPath, { mode: 0o700, recursive: true });
+    return dirPath;
   }
 
-  return `unix://${path.resolve(app.getPath('userData'), socketName)}`;
+  // On Linux the temp dir is like /tmp, and it's accessible by anyone,
+  // so create a user-specific subdirectory, e.g. "Teleport Connect-1000".
+  const dirPath = path.join(
+    appTempDir,
+    // UID can have at most 10 chars.
+    `${appName}-${process.getuid().toString()}`
+  );
+  await fs.promises.mkdir(dirPath, { mode: 0o700, recursive: true });
+  // The directory may already exist, validate it.
+  const stat = await fs.promises.lstat(dirPath);
```

Comment:
Instead of performing those checks below (which are hard errors if something is wrong) and caring about uid, what if we used `fs.mkdtemp` and created a new directory each time?

`fs.mkdtemp` doesn't let you specify the mode, but it's a thin wrapper on top of `mkdtemp`, which itself creates directories with mode 0700, according to `man mkdtemp`.

FWIW, we could have the same code for macOS and Linux then.

Based on https://www.netmeister.org/blog/mktemp.html, the only thing we're missing is cleaning up temporary files, but that is not super super important here because they're just sockets. We also wouldn't need to have the code which unlinks previous sockets.

Reply:
Oh, I was thinking that directories created with `fs.mkdtemp` inherit parent directory permissions, but you are right, they are created with 0700 mode.
I agree that it's much cleaner to use that API than to do all these checks by hand.

When it comes to cleaning up, I think we can skip it for now. I've left a TODO for myself :)